src/actions/pkg.cr

```diff
@@ -1,44 +1,150 @@
 module DppmRestApi::Actions::Pkg
   extend self
+  ALL_PKGS = ""
+  ONE_PKG  = "/:id"
+  include RouteHelpers
+
+  def build_config_response(package, builder)
+    builder.field package.package do
+      builder.object do
+        package.each_config_key do |key|
+          builder.field key, package.get_config key
+        end
+      end
+    end
+  end
+
+  def clean_unused_packages(context : HTTP::Server::Context)
+    if result = Actions.prefix.clean_unused_packages(confirmation: false) { }
+      return {data: result} if result.any?
+      bug = InternalServerError.new context, "received empty set from Prefix#clean_unused_packages; please report this strange bug"
+      error = NoPkgsToClean.new context
+      error.cause = bug
+      raise error
+    end
+    raise NoPkgsToClean.new context
+  end
+
   # List built packages
+  #
+  # TODO optional pagination
   relative_get nil do |context|
-    if context.current_user? && Actions.has_access? context, Access::Read
-      # TODO: list all built packages
+    if Actions.has_access? context, Access::Read
+      JSON.build context.response do |json|
+        json.object do
+          json.field "data" do
+            json.array do
+              Actions.prefix.each_pkg do |package|
+                json.object do
+                  json.field "package", package.package
+                  json.field "version", package.version
+                  package.each_config_key do |key|
+                    if parse_boolean_param key, from: context
+                      json.field key do
+                        package.get_config(key) { nil }.to_json json
```

Comment:
Use `get_config` instead, which will raise.
Note there is also `get_config?`

Reply:
> Use get_config instead, which will raise.

Note that the current implementation of `build_json` requires that [no exceptions be thrown](https://github.com/DFabric/dppm-rest-api/blob/adbae12b5e71c925ff0a1404631902849a4c3385/src/actions/route_helpers.cr#L42-L62) while the response is being built. Until now, the only resolution to this I could think of was building to an intermediary IO and copying to `context.response` after it's built. That would be a big performance hit in a crucial spot. However, talking another look at my implementation I was thinking that there is one alternative, although it is less than ideal....suggestions are welcome, but I'd recommend we discuss this in a separate PR, so that I can also begin work on other routes while we work out the details.

```crystal
def build_json(response : IO)
  error : Exception? = nil
  JSON.build response do |json|
    json.object do
      json.field "data" do
        json.object do
          yield json
        rescue e
          error = e
        end
      end
      if error
        json.field "errors" do
          ErrorResponse.new(error).errors.to_json json
        end
      end
    end
  end
end
```

I realize this would require specifying a partial succesful response, but there's no way to rewind an `HTTP::Server::Response` because it's *streamed* out to the client on the fly.

Reply:
For the package routes, I (think) I successfully avoided cases which raise, so that's an option as well, but it's...fragile, and I don't like it, especially not long-term

Reply:
Yes, we will have to figure out how to do it. We could have an abstraction for this and use it anywhere we need.
For a reason or another, building the JSON may result of an error.